### PR TITLE
do not recompute sidecar and gateway if not changed

### PR DESCRIPTION
**Pull request by ramaraochavali**
Do not recompute sidecar and merge gateways if they have not changed while pushing to proxy
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/xds/ads.go [hunk 1]
```diff
@@ -620,27 +623,52 @@ func (s *DiscoveryServer) initializeProxy(node *core.Node, con *Connection) erro
...
 		}
 	}
 }
 
-func (s *DiscoveryServer) setProxyState(proxy *model.Proxy, push *model.PushContext) {
-	proxy.SetWorkloadLabels(s.Env)
-	proxy.SetServiceInstances(push.ServiceDiscovery)
-
+func (s *DiscoveryServer) computeProxyState(proxy *model.Proxy, request *model.PushRequest) {
+	fmt.Println("computeProxyState.....")
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:639]: remove

File: pilot/pkg/xds/ads.go [hunk 2]
```diff
@@ -620,27 +623,52 @@ func (s *DiscoveryServer) initializeProxy(node *core.Node, con *Connection) erro
...
-	proxy.SetSidecarScope(push)
-	proxy.SetGatewaysForProxy(push)
+	var sidecar, gateway bool
+	push := s.globalPushContext()
+	if request == nil {
+		sidecar = true
+		gateway = true
+	} else {
+		push = request.Push
+		// This is mainly needed for tests.
```
Comment (howardjohn) [inline pilot/pkg/xds/ads.go:652]: its used for full pushes - mesh config change, network change, etc

File: pilot/pkg/xds/ads.go (hunk 1 above)
Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go]: done

File: pilot/pkg/xds/ads.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/xds/ads.go]: Ah .got it. removed the comment

**Review by adiprerepa — COMMENTED** [review]
LGTM but will yield to @howardjohn

**Review by howardjohn — APPROVED** [review]

**Pull request merged**